test/libsolidity/ASTJSONTest.cpp

```diff
@@ -141,13 +141,13 @@ TestCase::TestResult ASTJSONTest::run(ostream& _stream, string const& _linePrefi
 
 		if (!c.parseAndAnalyze(variant.stopAfter))
 		{
-			// Ignore non-fatal analysis errors, we only want to export.
-			if (c.state() > CompilerStack::State::Parsed)
-				continue;
-
-			SourceReferenceFormatter formatter(_stream, c, _formatted, false);
-			formatter.printErrorInformation(c.errors());
-			return TestResult::FatalError;
+			// We just want to export so raise fatal analysis errors only
+			if (c.state() < CompilerStack::State::ParsedAndImported)
```

Comment:
You only kind of invert the if-branches here, the logic basically stays the same (or did I miss something?).

After having a quick look, the problem seems to be that, if we call `parseAndAnalyze(AnalysisPerformed)` (one of the m_variants we have), the `c.state()` after it fails is `ParsedAndImported` which will not be caught by the `if` check here (and neither by the code before those changes).

So I believe the right change is to either change the `Parsed` in the old code to `ParsedAndImported` or to add change the check here to `<=`

Reply:
Actually, I made this change to get rid of "continue" instruction. It was responsible for skipping the rest of the test case (including "runTest" function call). Other changes are minor. Nonetheless, I can revert to "<= Parsed" if you feel that would be easier to read.

Reply:
Oooh I see. Good point. Yeah we should keep that change. I think the other parts of my comment are still valid though :)

Reply:
Yeah, I agree with @Marenz that it looks like it should be `<=`, not `<`. I.e. I see that for `not_existing_import.sol` we get both `not_existing_import.json` and `not_existing_import_parseOnly.json` but we should not be able to produce the former.

Reply:
> So we basically want to be able to parse partially broken code, so we can export it so that other tools can work with it.

But can we actually generate AST from code that does not even parse? It makes sense for code that parses but does not pass analysis but not sure it's true for code with parsing errors.

Reply:
> Yeah, I agree with @Marenz that it looks like it should be `<=`, not `<`. I.e. I see that for `not_existing_import.sol` we get both `not_existing_import.json` and `not_existing_import_parseOnly.json` but we should not be able to produce the former.

What would be a drawback of keeping "< ParsedAndImported" comparison? I assume it is expected to get AST generated for contracts with non-existing import files. Then I am not sure why to raise an error in tests with successful parsing and missing import files.

`not_existing_import.json` and `not_existing_import_parseOnly.json` are the same because the test ends up at compile step. I do not feel it is really bad. We at least make sure that AST is not updated in case of missing imports.

Reply:
The problem is that it works only for this specific case where there are errors in import resolution but not for errors at the analysis stage. I think that we should support both.

All the analysis errors you fixed look like genuine bugs in tests and not something meant to use that feature, so we're apparently missing coverage for this. I think you should add a new test meant specifically to test AST generation from code with analysis bugs.

Reply:
> But can we actually generate AST from code that does not even parse? It makes sense for code that parses but does not pass analysis but not sure it's true for code with parsing errors.

No no, as you said, for parse errors we always want to fail, but we do want to be able to export if parsing was successful even if there is an error in the analysis.

Reply:
ok. right.

I think it would be good to have some kind of flag saying if the file is supposed to have analysis errors. We did have quite a few bugs here and I guess the cases with intentional errors will be in the minority. We we should fail on analysis errors unless the test specifically tells us not to.

We could do it through file naming (easier but uglier and a bit too "magical" for my taste) or via comments in the file like we do in other isoltest test suites (nicer but might require adding some machinery here for parsing that).

Reply:
@cameel, won't we fail in AST comparison in case of analysis errors?

Reply:
It shouldn't. I'm assuming that generating AST in case of analysis errors does not crash of course - if it does, it's a separate matter and probably a bug to fix. But as long as  it does not, you should just get identical output and the comparison should pass.

Reply:
> It shouldn't. I'm assuming that generating AST in case of analysis errors does not crash of course - if it does, it's a separate matter and probably a bug to fix. But as long as it does not, you should just get identical output and the comparison should pass.

I did an investigation on that and found the analysis step extends AST by annotations. In result AST differs between ParsedAndImported and AnalysisPerformed states. Can't we rely on that?
Could you give me any solidity code that pass the parsing step and fails on the analysis so I can confirm we cover that case also?
I basically agree it would be nice to have clear information in test case if errors are expected. But I will complicate code and I am not sure if this is necessary.

Reply:
tests added